google-workloads-to-champsim/google-workload-to-champsim.cpp

```diff
@@ -0,0 +1,469 @@
+            memset(cs.source_registers, 0, sizeof(cs.source_registers));
+            memset(cs.source_memory, 0, sizeof(cs.source_memory));
+            sz = t.size;
+            address = t.addr;
+
+            break;
+
+        case TRACE_TYPE_INSTR_DIRECT_JUMP:
+        case TRACE_TYPE_INSTR_INDIRECT_JUMP:
+        case TRACE_TYPE_INSTR_CONDITIONAL_JUMP:
+        case TRACE_TYPE_INSTR_DIRECT_CALL:
+        case TRACE_TYPE_INSTR_INDIRECT_CALL:
+        case TRACE_TYPE_INSTR_RETURN:
+        case TRACE_TYPE_INSTR_SYSENTER:
+            // Branch Instr
+
+            if (is_first)
+            {
+                cs.ip = t.addr;
+                cs.is_branch = 1;
+                memset(cs.destination_registers, 0, sizeof(cs.destination_registers));
+                memset(cs.destination_memory, 0, sizeof(cs.destination_memory));
+                memset(cs.source_registers, 0, sizeof(cs.source_registers));
+                memset(cs.source_memory, 0, sizeof(cs.source_memory));
+                sz = t.size;
+                address = t.addr;
+                is_first = false;
+
+                if (t.type == TRACE_TYPE_INSTR_DIRECT_JUMP)
+                {
+                    cs.destination_registers[0] = REG_INSTRUCTION_POINTER; // writes_ip
+                }
+                else if (t.type == TRACE_TYPE_INSTR_INDIRECT_JUMP)
+                {
+                    cs.destination_registers[0] = REG_INSTRUCTION_POINTER; // writes_ip
+                    cs.source_registers[0] = 99;                           // reads_other
+                }
+                else if (t.type == TRACE_TYPE_INSTR_CONDITIONAL_JUMP)
+                {
+                    cs.source_registers[0] = REG_INSTRUCTION_POINTER;      // reads_ip
+                    cs.destination_registers[0] = REG_INSTRUCTION_POINTER; // writes_ip
+                    cs.source_registers[1] = REG_FLAGS;                    // reads_flag
+                }
+                else if (t.type == TRACE_TYPE_INSTR_DIRECT_CALL)
+                {
+                    cs.source_registers[0] = REG_STACK_POINTER;            // reads_sp
+                    cs.source_registers[1] = REG_INSTRUCTION_POINTER;      // reads_ip
+                    cs.destination_registers[0] = REG_STACK_POINTER;       // writes_sp
+                    cs.destination_registers[1] = REG_INSTRUCTION_POINTER; // writes_ip
+                }
+                else if (t.type == TRACE_TYPE_INSTR_INDIRECT_CALL)
+                {
+                    cs.source_registers[0] = REG_STACK_POINTER;            // reads_sp
+                    cs.source_registers[1] = REG_INSTRUCTION_POINTER;      // reads_ip
+                    cs.destination_registers[0] = REG_STACK_POINTER;       // writes_sp
+                    cs.destination_registers[1] = REG_INSTRUCTION_POINTER; // writes_ip
+                    cs.source_registers[2] = 99;                           // reads_other
```

Comment:
I'm concerned that instructions do not get any register information unless they're branches. It implies that there are no data dependencies in the trace, which is almost certainly not true. Furthermore, all indirect branches have data dependencies on each other (on register 99), which could lead to some strange results.

Reply:
I guess this is a limitation of the Google traces. Google traces do not have these info.

Reply:
If we do not have this information, I'm not sure we can move forward converting the Google traces to ChampSim traces. Memory timing is extremely important, and register information is the way that ChampSim generates timing.

Reply:
Regular drmemtrace traces have instruction encodings with this information so a converter is feasible there.  The version 1 release of the Google Workload traces had that information stripped out, and most core studies then had to perform limit studies assuming full deps vs zero deps.  We are working on a version 2 release of the Google workloads where we hope to get approval to release information on dependencies, though likely not full encodings.